### PR TITLE
XS✔ ◾ Added specification rule

### PR DESCRIPTION
**Pull request by danielmackay**
>
> 1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)

✏️ https://github.com/SSWConsulting/SSW.CleanArchitecture/issues/176

> 2. What was changed?

✏️ Added specification rule

> 3. Did you do pair or mob programming (list names)?

✏️ No

### CODE REVIEW

File: rules/use-specification-pattern/rule.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+   - common-design-patterns
+---
+
+When developing software, ensuring that your code is maintainable, flexible, and readable is crucial. One effective way to achieve this is by implementing the Specification pattern. This pattern allows for clear and modular encapsulation of business rules and query criteria, promoting the separation of concerns and enhancing the overall quality of your code.
+
+<!--endintro-->
+
+### What is the Specification Pattern?
+
+The Specification pattern is a design pattern used to define business rules in a reusable and combinable way. It encapsulates the logic of a business rule into a single unit, making it easy to test, reuse, and combine with other specifications.
```
Comment (matt-goldman-ssw) [inline rules/use-specification-pattern/rule.md:21]: I don't think combinable is a word is it? Would "composable" work here?

File: rules/use-specification-pattern/rule.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+### Use Cases for the Specification Pattern
+
+1. **Reusable Queries**: Specifications can be used to define reusable query criteria for data access layers, making it easier to build complex queries.
+2. **Validation Rules**: Specifications can be used to define validation rules for input data, ensuring that it meets the required criteria.
+3. **Encapsulating Business Rules**: Specifications can encapsulate complex business rules in the Domain where most business logic should go.
+4. **Repository Alternative**: Specifications can be used as an alternative to repositories for querying data.  Instead of encapsulating queries in repositories, you can encapsulate them in specifications.
+
+### Using the Specification Pattern
+
+Ardalis has created an excellent library called [Ardalis.Specifications](https://github.com/ardalis/Specification) that integrates well with EF Core.
```
Comment (matt-goldman-ssw) [inline rules/use-specification-pattern/rule.md:32]: I don't think "Ardalis" is clear here. Maybe Steve Smith, with a link to his online profile somewhere, or his website.

File: rules/use-specification-pattern/rule.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+2. **Use Specification**:
+
+    ```csharp
+        var teamId = new TeamId(request.TeamId);
+        var team = dbContext.Teams
+            .WithSpecification(new TeamByIdSpec(teamId))
+            .FirstOrDefault();
+    ```
+
+For an end-to-end example of the specification pattern see the [SSW.CleanArchitecture Template](https://github.com/SSWConsulting/SSW.CleanArchitecture).
```
Comment (matt-goldman-ssw) [inline rules/use-specification-pattern/rule.md:59]: To consider: link direct to the example in the template

**Review by matt-goldman-ssw — COMMENTED** [review]
A couple of comments to consider. Also I think this rule can be improved (can be in the future) using the "concrete before abstract" principle. It's a bit difficult to follow the conceptual path, whereas showing an example up front can help clarify what this pattern is and what it is used for. (As Adam likes to say "show the pain").

File: rules/use-specification-pattern/rule.md (hunk 2 above)
Comment (william-liebenberg) [inline rules/use-specification-pattern/rule.md:32]:
```suggestion
Steve Smith (aka "Ardalis") has created an excellent library called [Ardalis.Specifications](https://github.com/ardalis/Specification) that integrates well with EF Core.
```

Comment (danielmackay) [reply rules/use-specification-pattern/rule.md]: Updated.  Cheers for the suggestion.

Comment (danielmackay) [reply rules/use-specification-pattern/rule.md]: Done

File: rules/use-specification-pattern/rule.md (hunk 1 above)
Comment (danielmackay) [reply rules/use-specification-pattern/rule.md:21]: Done

File: rules/use-specification-pattern/rule.md (hunk 3 above)
Comment (danielmackay) [reply rules/use-specification-pattern/rule.md:59]: Added this PBI - https://github.com/SSWConsulting/SSW.CleanArchitecture/issues/338

**Comment by danielmackay** [comment]
> A couple of comments to consider. Also I think this rule can be improved (can be in the future) using the "concrete before abstract" principle. It's a bit difficult to follow the conceptual path, whereas showing an example up front can help clarify what this pattern is and what it is used for. (As Adam likes to say "show the pain").

Done

File: rules/use-specification-pattern/rule.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+## Good Example
+
+Re-visiting the example above, we can apply the Specification pattern as follows:
+
+```csharp
+public sealed class OrderByIdSpec : SingleResultSpecification<Order>
+{
+   public OrderByIdSpec(Guid orderId)
```
Comment (matt-goldman-ssw) [inline rules/use-specification-pattern/rule.md:108]:
```suggestion
   public IncompleteOrderByIdSpec(Guid orderId)
```

File: rules/use-specification-pattern/rule.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            .Where(o => o.Id == orderId && o.Status != OrderStatus.Complete);
+   }
+} 
+
+public class OrderService(ApplicationDbContext dbContext)
+{
+    public async Task AddItem(Guid orderId, OrderItem item)
+    {
+        var order = dbContext.Orders
+            .WithSpecification(new OrderByIdSpec(orderIdorderId))
```
Comment (matt-goldman-ssw) [inline rules/use-specification-pattern/rule.md:121]:
```suggestion
            .WithSpecification(new IncompleteOrderByIdSpec(orderIdorderId))
```

File: rules/use-specification-pattern/rule.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        
+        order.AddItem(item);
+        
+        await dbContext.SaveChangesAsync();
+    }
+    
+    public void RemoveItem(int age)
+    {
+        var order = dbContext.Orders
+            .WithSpecification(new OrderByIdSpec(orderIdorderId))
```
Comment (matt-goldman-ssw) [inline rules/use-specification-pattern/rule.md:132]:
```suggestion
            .WithSpecification(new IncompleteOrderByIdSpec(orderIdorderId))
```

**Review by matt-goldman-ssw — APPROVED** [review]
LGTM

File: rules/use-specification-pattern/rule.md (hunk 5 above)
Comment (danielmackay) [reply rules/use-specification-pattern/rule.md:121]: Done

**Review by william-liebenberg — APPROVED** [review]

**Pull request merged**